### PR TITLE
APEXCORE-575 Improve application restart time.

### PR DESCRIPTION
**Pull request by tushargosavi**
I have implemented a new storage agent (CascadeStorageAgent) which maintains two storage agents one for old checkpoint directory and one for new checkpoint directory, using this storage agent we could direct read on old directory during initial start, and write to new checkpoint directory. With this we could avoid copy of checkpoints from old directory to new directory. with 2 GB state application restart was brought down to few seconds from 2 minutes.

Other changes are
- Add log message to print time taken to copy the initial state.
- Do not copy stats and events directory as they are overwritten anyway in the new application.
- Use new storage agent to avoid copying checkpoints directory.

### CODE REVIEW

**Comment by sandeshh** [comment]
Should the new StorageAgent become a part of Malhar?

**Comment by tushargosavi** [comment]
This will introduce circular dependency on malhar, as this StorageAgent is used by default during application relaunch. Also application may not have dependency on malhar if all operators are custom. this is same as default storage agents being in apex (FSStorageAgent)

**Comment by tushargosavi** [comment]
@PramodSSImmaneni @vrozov can you review this pull request?

**Comment by bhupeshchawda** [comment]
@tushargosavi the branch has conflicts. Can you please rebase?

**Comment by tushargosavi** [comment]
@bhupeshchawda I have updated branch and resolved conflicts.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java [hunk 1]
```diff
@@ -94,4 +94,10 @@
     public void setApplicationAttributes(AttributeMap map);
   }
 
+  interface AsyncStorageAgent
```
Comment (bhupeshchawda) [inline api/src/main/java/com/datatorrent/api/StorageAgent.java]: Add documentation for interface and methods

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import com.datatorrent.api.StorageAgent;
+import com.datatorrent.api.StorageAgent.AsyncStorageAgent;
+
+/**
+ *
+ */
+public class CascadeStorageAgent implements StorageAgent, AsyncStorageAgent, Serializable
+{
+  private static final Logger logger = LoggerFactory.getLogger(CascadeStorageAgent.class);
+  private StorageAgent parent;
```
Comment (bhupeshchawda) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: declare final?

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  private StorageAgent current;
+  private transient Map<Integer, long[]> oldOperatorToWindowIdsMap = Maps.newHashMap();
+
+  public CascadeStorageAgent(StorageAgent parent, StorageAgent current)
+  {
+    this.parent = parent;
+    this.current = current;
+    oldOperatorToWindowIdsMap = Maps.newHashMap();
```
Comment (bhupeshchawda) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Already initialized in the declaration

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+   *
+   * @param operatorId
+   * @return
+   * @throws IOException
+   */
+  private long[] getOldWindowIds(int operatorId) throws IOException
+  {
+    long[] oldWindowIds = oldOperatorToWindowIdsMap.get(operatorId);
+    if (oldWindowIds == null) {
+      oldWindowIds = parent.getWindowIds(operatorId);
```
Comment (bhupeshchawda) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: any chance `parent.getWindowIds` returns null?

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 5]
```diff
@@ -3192,23 +3193,43 @@ public void setApplicationId(LogicalPlan newApp, Configuration conf)
...
-        FSStorageAgent fssa = (FSStorageAgent)sa;
-        if (fssa.path.contains(oldAppId)) {
-          fssa = new FSStorageAgent(fssa.path.replace(oldAppId, appId), conf);
-          lp.setAttribute(OperatorContext.STORAGE_AGENT, fssa);
-        }
+      lp.setAttribute(OperatorContext.STORAGE_AGENT, updateStorageAgent(sa, oldAppId, appId, conf));
+    }
+  }
+
+  static StorageAgent updateStorageAgent(StorageAgent sa, String oldAppId, String appId, Configuration conf)
```
Comment (bhupeshchawda) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: private?

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+   * @throws IOException
+   */
+  @Override
+  public void delete(int operatorId, long windowId) throws IOException
+  {
+    current.delete(operatorId, windowId);
+    // cleanup windowIds from cached data.
+    long[] oldWindowIds = getOldWindowIds(operatorId);
+    if (oldWindowIds.length != 0) {
+      oldOperatorToWindowIdsMap.put(operatorId, new long[0]);
```
Comment (bhupeshchawda) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Check if `oldOperatorToWindowIdsMap` can be removed.
Can you add a case for delete in the tests as well?

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 1 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: done

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 2 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: done

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 3 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Removed oldOperatorToWindowIdsMap as suggested in comment further down.

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 4 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: There is no contract for return value from getWindowIds. Existing implementations throws an exception if there are no windowIds available or zero size array. I will add a null check here too.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 5 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: done

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 6 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Cache is needed at some place to check if a checkpoint belong to parent or not. without cache we will need to access filesystem to get the list of checkpoits.

**Comment by tushargosavi** [comment]
@bhupeshchawda I have addressed the comments , please check again.

**Review by bhupeshchawda — APPROVED** [review]
Looks good to me.
Will merge this soon if there are no other comments from the community.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java [hunk 7]
```diff
@@ -94,4 +94,24 @@
...
+     */
+    void finalize(int operatorId, long windowId) throws IOException;
+
+    /**
+     * Check if StorageAgent is configured to take synchronous checkpoints.
+     *
+     * @return true if StorageAgent is configured to take synchronous checkpoints.
+     * @return false otherwise.
+     */
+    boolean isSyncCheckpoint();
```
Comment (pramodin) [inline api/src/main/java/com/datatorrent/api/StorageAgent.java]: Shouldn't this method be in the base interface?

File: api/src/main/java/com/datatorrent/api/StorageAgent.java [hunk 8]
```diff
@@ -94,4 +94,24 @@
...
+  {
+    /**
+     * Make checkpoint for given windowID final. i.e after this method returns,
+     * the checkpoint is accessible for recovery.
+     *
+     * @param operatorId
+     * @param windowId
+     * @throws IOException
+     */
+    void finalize(int operatorId, long windowId) throws IOException;
```
Comment (pramodin) [inline api/src/main/java/com/datatorrent/api/StorageAgent.java]: Is this a way to write the checkpoint immediately in some scenarios? Why call it finalize? It should probably be something like saveImmediately.

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    long[] oldWindowIds = getParentWindowIds(operatorId);
+    if (oldWindowIds.length >= 1 && windowId <= oldWindowIds[oldWindowIds.length - 1]) {
+      return parent.load(operatorId, windowId);
+    }
+    return current.load(operatorId, windowId);
+  }
+
+  @Override
+  public long[] getWindowIds(int operatorId) throws IOException
+  {
```
Comment (pramodin) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Shouldn't this return a collection of both sets of window ids and also use deleted window id as a starting cutoff for the old window ids. For example, the first operator may have checkpointed in the new run but none of the other operators. If a committed window calculation is happening, there is a still a committed window in the old checkpoints.

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+   * storage agent. and cached state for old storage agent is removed.
+   *
+   * @param operatorId
+   * @param windowId
+   * @throws IOException
+   */
+  @Override
+  public void delete(int operatorId, long windowId) throws IOException
+  {
+    if (!isCheckpointFromParent(operatorId, windowId)) {
```
Comment (pramodin) [inline common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Might need to store deleted window id and use it as a cutoff to return old checkpoints. See comment on getWindowIds below.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 11]
```diff
@@ -3192,23 +3193,43 @@ public void setApplicationId(LogicalPlan newApp, Configuration conf)
...
+      lp.setAttribute(OperatorContext.STORAGE_AGENT, updateStorageAgent(sa, oldAppId, appId, conf));
+    }
+  }
+
+  private static StorageAgent updateStorageAgent(StorageAgent sa, String oldAppId, String appId, Configuration conf)
+  {
+    if (sa instanceof AsyncFSStorageAgent || sa instanceof FSStorageAgent) {
+      FSStorageAgent newAgent = (FSStorageAgent)updateFSStorageAgent(sa, oldAppId, appId, conf);
+      if (newAgent != sa) {
+        return new CascadeStorageAgent(sa, newAgent);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:3251]: Why not set cascade storage agent only when application is being recovered from a prior run, otherwise it is an extra layer of indirection.

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:3251]: This is called from StramClient when we re-write recovery file from old application. hence it will be set when application is restarted from prior run.

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 9 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: true, we need to return windowIds from both the agents. doing this change. But we don't need to cutoff at deleted windowIds. checkpoint deletion is an optimization to reduce the disk space. the recovery window id calculation and committed windowId calculation will work correctly without deletion of checkpoints.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 8 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: The only places where async agent writes checkpoint immediately is while launching the application, where initial checkpoints are written by master. and while running tests. the name could be anything the intent of this method is the checkpoints are available for recovery after this method returns. If I rename this to saveImmediately, then it seems similar to save, but don't have same parameters as save.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 7 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: We can't change the base interface as it would make it backward incompatible. and existing StorageAgent needs to be modified (Geode storage agent form Malhar).

File: common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java (hunk 10 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/CascadeStorageAgent.java]: Delete cutoff is not needed, getWindowIds is called during stram initialization for recovery window and committed window calculations. they will work correctly with extra windowIds present.
The another instance where getWindowIds is called is during recovery with atmost-once and exactly once mode, where the max(getWindowIds) is used hence that part will be correct too as thee will be an checkpointed windowId > deletedWindowId.

**Comment by tushargosavi** [comment]
@PramodSSImmaneni  I have handled the comments, can you please review the changes.

**Comment by pramodin** [comment]
@sandeshh will look into it

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 7 above)
Comment (pramodin) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: On second thoughts.. it looks like the sync checkpoint flag is to force a sync behavior on the async agent and is being used by tests. I don't think this should be part of the interface. Second, I am not sure if this functionality is needed as the tests seem to be creating the AsyncFSStorageAgent and setting this flag immediately, in that case why not just create a FSStorageAgent instead?  If this functionality is needed for testing, it can be an internal artifice. One way to do this, for example, is to define a new internal interface, not in api, and add this method there. You could call this interface SyncCompatibleAgent from example.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java [hunk 12]
```diff
@@ -94,4 +96,31 @@
     public void setApplicationAttributes(AttributeMap map);
   }
 
+  /**
+   * Storage agent which can take checkpoints asynchronously.
```
Comment (pramodin) [inline api/src/main/java/com/datatorrent/api/StorageAgent.java]: It is not clear what someone who wants to implement a new async storage agent implementation needs to do. I think to start with, it is confusing that they would need to implement both the StorageAgent and the AsyncStorageAgent interface. Then it needs to be documented that they need to the operation in two steps. First, a store method where they need to make a temporary copy of the given object and then a finalize method where they can store it.

The same store method seems to have two different meanings when it comes to sync case vs async case. In sync case it is supposed to completely save the checkpoint. In async case it is supposed to only make a temporary copy. This isn't the correct way to expose it to the user either.

I suggest for these reasons to not have this interface now and rethink this interface in the future.

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 8 above)
Comment (pramodin) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: Finalize sounds good.

File: common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java [hunk 13]
```diff
@@ -153,6 +154,16 @@ public void delete(int operatorId, long windowId) throws IOException
   public long[] getWindowIds(int operatorId) throws IOException
   {
     Path lPath = new Path(path + Path.SEPARATOR + String.valueOf(operatorId));
+    try {
```
Comment (pramodin) [inline common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:157]: Seems to be unrelated to the main intent of this PR. Is this a fix for a different bug?

Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:157]: during application restart as the new directory for current application is not created, the getWindowIds will throw an exception. hence this try is added to catch FileNotFoundException

File: api/src/main/java/com/datatorrent/api/StorageAgent.java (hunk 12 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StorageAgent.java]: Making this interface private and not part of API.

**Comment by tushargosavi** [comment]
@PramodSSImmaneni I have updated the pull request please check.

File: common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+{
+  /**
+   * Make checkpoint for given windowID final. i.e after this method returns,
+   * the checkpoint is accessible for recovery.
+   *
+   * @param operatorId
+   * @param windowId
+   * @throws IOException
+   */
+  public void finalize(int operatorId, long windowId) throws IOException;
```
Comment (vrozov) [inline common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java:44]: Can it be changed to sync() or flush() instead of finalize() that has special meaning in Java?

File: common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java (hunk 13 above)
Comment (vrozov) [reply common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:157]: Will it be better to try to create the directory if it does not exist and continue with the normal flow?

File: common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java [hunk 15]
```diff
@@ -153,6 +154,16 @@ public void delete(int operatorId, long windowId) throws IOException
...
+      FileStatus status = fileContext.getFileStatus(lPath);
+      if (!status.isDirectory()) {
+        throw new IOException("Checkpoint location is not a directory ");
```
Comment (vrozov) [inline common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:160]: What if this is a link to a directory? Why is this an IOException and not a Runtime? What caller is supposed to do with it? Remove extra space at the end of the message.

File: common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+  /**
+   * Check if StorageAgent is configured to take synchronous checkpoints.
+   *
+   * @return true if StorageAgent is configured to take synchronous checkpoints.
+   * @return false otherwise.
+   */
+  public boolean isSyncCheckpoint();
```
Comment (vrozov) [inline common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java:52]: Will it be better to change the condition? On AsyncStorageAgent it will be more natural to ask for Async.

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+
+/**
+ * A StorageAgent which chains two StorageAgent. It use the current storage-agent to store
+ * the checkpoint, and use the parent agent to read old checkpoints. For application having
+ * large number of physical operators, the size and number of files to be copied could be
+ * large impacting application restart time. This storage-agent is used during application
+ * restart to avoiding copying checkpoints from old application directory to improve application
+ * restart time.
+ */
+public class CascadeStorageAgent implements StorageAgent, AsyncStorageAgent, Serializable
```
Comment (vrozov) [inline common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:42]: Why is this class in common? Can it be moved to the engine?

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+{
+  private static final Logger logger = LoggerFactory.getLogger(CascadeStorageAgent.class);
```
Comment (vrozov) [inline common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:44]: serialVersionUID?

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+  @Override
+  public boolean isSyncCheckpoint()
+  {
+    if (parent instanceof AsyncStorageAgent) {
+      return ((AsyncStorageAgent)parent).isSyncCheckpoint();
+    }
+    return true;
+  }
+
+  public Object readResolve() throws ObjectStreamException
```
Comment (vrozov) [inline common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:188]: Why readResolve?

**Review by vrozov — COMMENTED** [review]
Sorry for late comments, but possibly they can be addressed in a separate PR.

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java (hunk 18 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:44]: will add it.

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java (hunk 19 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:188]: @vrozov can I make these changes as part of different JIRA?

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java (hunk 17 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:42]: Other storage agents were in common, hence kept this storage agent in the same module.

Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:42]: it can be moved to engine.

File: common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java (hunk 14 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java:44]: will change it to flush.

File: common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java (hunk 16 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java:52]: will make the change.

File: common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java (hunk 19 above)
Comment (vrozov) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:188]: Yes, it can be part of a different JIRA.

Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/CascadeStorageAgent.java:188]: read resolve was to initialize a final transient variable through constructor.  removed final now and initialized the variable in readObject.

File: common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java (hunk 15 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:160]: Removed extra space and changed to RuntimeException.

File: common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java (hunk 13 above)
Comment (tushargosavi) [reply common/src/main/java/com/datatorrent/common/util/FSStorageAgent.java:157]: will do the change.

File: common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java (hunk 16 above)
Comment (tushargosavi) [reply common/src/main/java/org/apache/apex/common/util/AsyncStorageAgent.java:52]: On second though, AsyncStorageAgent is expected to work in async mode by default, hence code using it can assume that behavior, only when the code needs to check for explicit sync configuration it can use this method. hence the current name looks natural to me.

**Comment by tushargosavi** [comment]
@vrozov opened JIRA APEXCORE-709 for it. will be opening a pull request shortly.

**Comment by tushargosavi** [comment]
@vrozov I have opened pull request https://github.com/apache/apex-core/pull/517 for recent suggestions.
